ArchitectureSpec.md

```diff
@@ -0,0 +1,74 @@
+# Vision Input Team:
+Handles image, video, and camera input and generates arrays of RGB or depth data
+## Responsibilities
+- Finding Camera parameters
+- Field of View (FoV)
+- Handle multiple cameras
+- Synchronization
+## Class structure
+4 ImageGenerator classes
+- `DepthDataGenerator (single frame depth filename, still image filename)`
+- `DepthLiveGenerator (Realsense camera port int)`
+- `ImageGenerator (single image frame filename)`
+- `VideoFileGenerator(filename)`
+- `VideoLiveGenerator(camera port input int)`
+## API
+### get_horizontal_fov()
+Returns float degrees
+
+### get_vertical_fov()
+Returns float degrees
+
+### generate()
+Gets data for one frame
+
+Returns: `tuple(rgb ndarray shape (x, y, 3), depth ndarray shape(x,y))`
+For sources with no depth data, return `tuple(ndarray, None)`
+
+# Detection Team:
+Accurately detect any required target on the field
+## Responsibilities
+- Identify pixels that correspond to target (goal hexagon, balls)
+- Construct contours around targets (goal hexagon)
+- Color calibration to correct for lighting changes
+## Class structure
+1 class for each of 3 different targets. Constructor takes in a ImageGenerator
+- `LoadingBayDetector`
+- `PowerPortDetector`
+- `BallDetector`
+## API
+### get_generator()
+Returns the contained ImageGenerator
+
+### run_detection()
+Calls ImageGenerator.generate()
+
+Returns: List of contours, one per target sorted by enclosed area
+
+"Contour" is output of [cv2.findContours](https://docs.opencv.org/2.4/modules/imgproc/doc/structural_analysis_and_shape_descriptors.html#findcontours)
+- For goal, return list with one contour (U shape)
+- For loading, return list with two contours (inside and outside)
+- For balls, return list with zero or more contours 
+
+# Situational Awareness: 
+Utilizes contours from detection team to gain actionable data for the robot
+## Responsibilities
+- Position of camera relative to targets (3D position)
+- Angle (3D)
+## Class Structure
+Constructor takes in a *Detector object
+
+## API
+### run_cv()
+Call Detector.run_detection()
+
+Return: `tuple(Euler Rotation: (yaw, pitch, roll) in degrees, World Position (x, y, z) in meters)`
```

Comment:
We need to also provide another angle value for the relative heading of the target.

Reply:
Hm, mathematically there's only one set of translation and rotation. However you're right that we haven't defined what the reference point is.

We could define the camera position to be at the origin and the rotation to be zero. Then we would be returning the relative position of the target and the rotation of the target.

Or we could define the target to be at the origin and its rotation to be zero. Then we would be returning the relative position of the camera and the rotation of the camera.

Reply:
Would defining the camera position to be at the origin be more consistent across all use cases? That way it may be easier for us to compare values when testing. However, is there any difference in calculations necessary between the two possible reference points, when it comes to the challenge?

Reply:
It's basically just a sign change. Or matrix inverse if you're thinking in terms of transform matrices.